marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/ConstraintsSpec.hs

```diff
@@ -0,0 +1,178 @@
+{-# LANGUAGE DataKinds #-}
+
+module Language.Marlowe.Runtime.Transaction.ConstraintsSpec
+  ( spec
+  ) where
+
+import qualified Cardano.Api as C
+import qualified Cardano.Api.Shelley as C
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import qualified Data.Map as Map
+import Data.Monoid (First(..), getFirst)
+import Data.Ratio ((%))
+import qualified Data.Set as Set
+import Gen.Cardano.Api.Typed
+import Language.Marlowe.Runtime.Cardano.Api (fromCardanoAddressAny, fromCardanoAddressInEra)
+import qualified Language.Marlowe.Runtime.ChainSync.Api as Chain
+import qualified Language.Marlowe.Runtime.Core.Api as Core.Api
+import Language.Marlowe.Runtime.Transaction.Constraints (TxConstraints(..))
+import qualified Language.Marlowe.Runtime.Transaction.Constraints as Constraints
+import qualified Language.Marlowe.Runtime.Transaction.Constraints as TxConstraints
+import Test.Hspec (Spec, expectationFailure, shouldBe)
+import qualified Test.Hspec as Hspec
+import qualified Test.Hspec.QuickCheck as Hspec.QuickCheck
+import qualified Test.QuickCheck as QuickCheck
+import Test.QuickCheck.Hedgehog (hedgehog)
+
+byteStringGen :: QuickCheck.Gen ByteString
+byteStringGen = BS.pack <$> QuickCheck.arbitrary
+
+emptyTxConstraints :: TxConstraints 'Core.Api.V1
+emptyTxConstraints = TxConstraints
+  { marloweInputConstraints = TxConstraints.MarloweInputConstraintsNone
+  , payoutInputConstraints = Set.empty
+  , roleTokenConstraints = TxConstraints.SpendRoleTokens Set.empty
+  , payToAddresses = Map.empty
+  , payToRoles = Map.empty
+  , marloweOutputConstraints = TxConstraints.MarloweOutputConstraintsNone
+  , merkleizedContinuationsConstraints = mempty
+  , signatureConstraints = Set.empty
+  , metadataConstraints = Map.empty
+  }
+
+spec :: Spec
+spec = do
+  Hspec.QuickCheck.prop "mustSpendToken" do
+    policyId <- Chain.PolicyId <$> byteStringGen
+    tokenName <- Chain.TokenName <$> byteStringGen
+
+    let
+      assetId :: Chain.AssetId
+      assetId = Chain.AssetId policyId tokenName
+
+      actual :: Constraints.TxConstraints 'Core.Api.V1
+      actual = Constraints.mustSpendRoleToken assetId
+
+      expected :: Constraints.TxConstraints 'Core.Api.V1
+      expected = emptyTxConstraints
+        { roleTokenConstraints = Constraints.SpendRoleTokens $ Set.singleton assetId }
+
+    pure $ actual `shouldBe` expected
+
+  Hspec.describe "adjustTxForMinUtxo" do
+
+    Hspec.QuickCheck.prop "The Marlowe output has NOT been adjusted" do
+      protocol <- hedgehog genProtocolParameters
```

Comment:
How come this test uses `genProtocolParameters` while all the others use the hard-coded one?

Reply:
The one test runs fine with generated protocol parameters but the others do not. Truly arbitrary protocol parameters as created by `Gen.Cardano.Api.Typed` break the other tests with errors like this below.

```
1) Language.Marlowe.Runtime.Transaction.Constraints.adjustTxForMinUtxo all outputs satisfy Cardano API minimum UTxO requirements
       uncaught exception: ErrorCall
       toBabbagePParams: must specify protocolParamUTxOCostPerByte
       CallStack (from HasCallStack):
         error, called at src/Cardano/Api/ProtocolParameters.hs:1588:3 in cardano-api-1.35.3-ad328aa7763ad4ae1b552540147c8763f3caf0c330b6cc738756de67f73c32a3:Cardano.Api.ProtocolParameters
       (after 1 test)
```

Reply:
With the changes to fix the Marlowe address output test, we are now using the hard-coded protocol parameters in all of these tests.